eth/db/storage.py

```diff
@@ -0,0 +1,477 @@
+    #
+    # Access & Modify Storage API
+    #
+    @abstractmethod
+    def get(self, address: Address, slot: int, from_journal: bool=True) -> int:
+        pass
+
+    @abstractmethod
+    def set(self, address: Address, slot: int, value: int) -> None:
+        pass
+
+    @abstractmethod
+    def delete(self, address: Address) -> None:
+        """
+        Delete *all* storage values in this account. This action is journaled, like set() actions.
+
+        Unlike set(), deleting storage will not cause :meth:`make_storage_roots` to emit a new
+        storage root (and therefore will not persist deletes to the base database).
+        The account's storage root must be explicitly set to the empty root.
+        """
+        pass
+
+    #
+    # Persistance API
+    #
+    @abstractmethod
+    def persist(self) -> Tuple[Tuple[Address, Hash32], ...]:
+        """
+        Push all pending changes into the database. This can only be called if there are no pending
+        changes since the last time :meth:`make_storage_roots` was called. Otherwise, a
+        ValidationError will be raised.
+
+        :return: the addresses and storage roots that were updated during storage persistance
+        """
+        pass
+
+    @abstractmethod
+    def make_storage_roots(self) -> Tuple[Tuple[Address, Hash32], ...]:
+        """
+        This generates the storage root for every pending change, and returns the addresses of
+        accounts that were changed, and the new storage root for that account.
+
+        This is an expensive operation, so should be called as little as possible. For example,
+        pre-Byzantium, this is called after every transaction, because we need the state root
+        in each receipt. Byzantium+, we only need state roots at the end of the block,
+        so we *only* call it right before persistance.
+
+        :return: the addresses and storage roots that will be updated at storage persistance
+        """
+        pass
+
+    #
+    # Record and discard Checkpoints API
+    #
+    @abstractmethod
+    def record(self) -> UUID:
+        pass
+
+    @abstractmethod
+    def discard(self, changeset: UUID) -> None:
+        pass
+
+    @abstractmethod
+    def commit(self, changeset: UUID) -> None:
+        pass
+
+
+class StorageLookup(BaseDB):
+    """
+    This lookup converts lookups of storage slot integers into the appropriate trie lookup.
+    Similarly, it persists changes to the appropriate trie at write time.
+
+    StorageLookup also tracks the state roots changed since the last persist.
+    """
+    logger = cast(ExtendedDebugLogger, logging.getLogger("eth.db.storage.StorageLookup"))
+
+    def __init__(self, db: BaseDB, account_db: BaseAccountDB, address: Address) -> None:
+        self._db = db
+        self._account_db = account_db
+        self._address = address
+        self._write_trie = None
+        self._starting_root_hash = None  # type: Hash32
+        self._trie_nodes_batch = None  # type: BatchDB
```

Comment:
Creating a `BatchDB` is a pretty lightweight operation, what do you think of simplifying the code by creating it here? As-is, you'll cause an exception if you call `commit_to()` too soon.

Reply:
It's not like a race condition or anything, it would just be wrong to call `commit` when there are no pending writes. I'll update the docstring and raise an explicit `ValidationError`.

Reply:
Yeah, it's definitely not a race-condition! And certainly the reason why this bug exists is that users aren't expected to call `commit_to()` with no changes. But, I'd still argue that it's a bug: there's nothing really wrong with calling `commit_to()` when there are no changes. My theory is that if `BatchDB` always exists then this class of bug is impossible in future code.

I'm also kind of surprised that mypy doesn't catch this. Even if I change the type annotation to `Optional[BatchDB]` it still doesn't complain about the null dereference :cry: